ballerina/soap_utils.bal

```diff
@@ -1,106 +1,186 @@
+//
+// WSO2 LLC. licenses this file to you under the Apache License,
+// Version 2.0 (the "License"); you may not use this file except
+// in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+import soap.wssec;
+
+import ballerina/crypto;
+import ballerina/http;
+import ballerina/mime;
+import ballerina/lang.regexp;
+import ballerina/jballerina.java;
+
+public isolated function validateTransportBindingPolicy(ClientConfig config) returns Error? {
+    if config.httpConfig.secureSocket is () {
+        wssec:InboundSecurityConfig|wssec:InboundSecurityConfig[] securityPolicy = config.inboundSecurity;
+        if securityPolicy is wssec:TransportBindingConfig {
+            return error Error(INVALID_PROTOCOL_ERROR);
+        } else if securityPolicy is wssec:InboundSecurityConfig[] {
+            foreach wssec:InboundSecurityConfig policy in securityPolicy {
+                if policy is wssec:TransportBindingConfig {
+                    return error Error(INVALID_PROTOCOL_ERROR);
+                }
+            }
+        }
+    }
+}
+
+public isolated function getReadOnlyRecords(ClientConfig original) returns readonly & ClientConfig = @java:Method {
+        'class: "org.wssec.WsSecurity"
+} external;
+
+public isolated function applySecurityPolicies(wssec:InboundSecurityConfig|wssec:InboundSecurityConfig[] inboundSecurity,
+                                      xml envelope) returns xml|wssec:Error {
+    wssec:InboundSecurityConfig|wssec:InboundSecurityConfig[] securityPolicy = inboundSecurity;
+    xml securedEnvelope;
+    if securityPolicy is wssec:InboundSecurityConfig {
+        if securityPolicy is wssec:TimestampTokenConfig {
+            securedEnvelope = check wssec:applyTimestampToken(envelope, securityPolicy);
+        } else if securityPolicy is wssec:UsernameTokenConfig {
+            securedEnvelope = check wssec:applyUsernameToken(envelope, securityPolicy);
+        } else if securityPolicy is wssec:SymmetricBindingConfig {
+            securedEnvelope = check wssec:applySymmetricBinding(envelope, securityPolicy);
+        } else if securityPolicy is wssec:AsymmetricBindingConfig {
+            securedEnvelope = check wssec:applyAsymmetricBinding(envelope, securityPolicy);
+        } else {
+            securedEnvelope = envelope;
+        }
+    } else {
+        foreach wssec:InboundSecurityConfig policy in securityPolicy {
+            securedEnvelope = check applySecurityPolicies(policy, envelope);
+        }
+    }
+    return securedEnvelope;
+}

```

Comment:
```suggestion
public isolated function applySecurityPolicies(wssec:InboundSecurityConfig|wssec:InboundSecurityConfig[] inboundSecurity,
                                      xml envelope) returns xml|wssec:Error {
    if inboundSecurity is wssec:TimestampTokenConfig {
        return wssec:applyTimestampToken(envelope, inboundSecurity);
    } else if inboundSecurity is wssec:UsernameTokenConfig {
        return wssec:applyUsernameToken(envelope, inboundSecurity);
    } else if inboundSecurity is wssec:SymmetricBindingConfig {
        return wssec:applySymmetricBinding(envelope, inboundSecurity);
    } else if inboundSecurity is wssec:AsymmetricBindingConfig {
        return wssec:applyAsymmetricBinding(envelope, inboundSecurity);
    } else if inboundSecurity is wssec:InboundSecurityConfig {
        return envelope;
    } 
    
    xml securedEnvelope;
    foreach wssec:InboundSecurityConfig policy in inboundSecurity {
        securedEnvelope = check applySecurityPolicies(policy, envelope);
    }
    return securedEnvelope;
}
```

can't we do something like this ?

Reply:
Fixed in [e309edb](https://github.com/ballerina-platform/module-ballerina-soap/pull/103/commits/e309edbbc4d7673f4c2ece8673433cf028487275)